 ### 1 - Carregar os arquivos e cadastro, investimento1 e investimento2 e mostrar a quantidade de linhas de cada um, além das informações desses dataframes

In [68]:
# Importa as bibliotecas utilizadas
!pip install pandas
import pandas as pd

Defaulting to user installation because normal site-packages is not writeable


In [69]:
# Carrega arquivo de cadastro
cadastro = pd.read_csv('cadastro.csv')

# Print do número de linhas e colunas do dataframe
print('Linhas', cadastro.shape[0])
print('Colunas', cadastro.shape[1])

# Primeiras cinco linhas do dataframe
cadastro.head()


Linhas 38754
Colunas 9


id                              Nome  Sexo        Nasc  Idade  \
0  id0000001             Eunice Hellen do Vale   Fem  1970-11-12     51   
1  id0000002   Sebastião Castelo Branco Júnior  Masc  2019-12-10      2   
2  id0000003         Romildo Fábio de Carvalho  Masc  2015-12-28      6   
3  id0000004         Lana Albina de Lemos Mumy   Fem  1993-05-03     28   
4  id0000005  Rose Elisângela Vyrnium de Moura   Fem  1984-06-10     37   

                     Email                                         Lat_long  \
0         djones@yahoo.com   (Decimal('24.7024355'), Decimal('-28.038258'))   
1   barnesandrea@yahoo.com   (Decimal('21.8589875'), Decimal('122.830566'))   
2  walkermichael@yahoo.com   (Decimal('-28.2161525'), Decimal('97.140529'))   
3   cobbangela@hotmail.com   (Decimal('-40.8557635'), Decimal('-2.936099'))   
4  crawfordcraig@yahoo.com  (Decimal('-20.2620885'), Decimal('104.349223'))   

                                     Job Estado  
0                       Technical author     TO  
1                   Biochemist, clinical     DF  
2  Development worker, international aid     DF  
3              Estate manager/land agent     DF  
4    Chartered public finance accountant     MS

In [70]:
# Carrega arquivo investimento_parte1
fin_01 = pd.read_csv('investimento_parte1.csv')

# Print do número de linhas e colunas do dataframe
print('Linhas', fin_01.shape[0])
print('Colunas', fin_01.shape[1])

# Primeiras cinco linhas do dataframe
fin_01.head()

Linhas 20000
Colunas 12


id  Salario       Perfil  Ativo01  Ativo02  Ativo03   Ativo04  \
0  id0011529    $2.25    Agressivo  1311.38      NaN      NaN  29267.48   
1  id0030644   $24.57    Agressivo  1007.38  4846.66  2717.82  12532.15   
2  id0019311  $367.53     Moderado  1128.99  4668.94  3092.60   9777.77   
3  id0036705   $23.26     Moderado  1185.26  4782.17  2669.05  14047.87   
4  id0001368    $6.31  Conservador   957.64  5585.52      NaN   5067.08   

   Ativo05  Ativo06  Ativo07      Nota1      Nota2  
0  4993.30      NaN   506.40  73.007216  30.025005  
1  4282.98      NaN   499.93  20.000123  29.477728  
2  4246.30  2007.83   509.50  24.901450  49.659852  
3  6026.84  2064.09   489.82  14.246718  29.271206  
4  4030.51  2022.61   507.39   5.922253  11.840601

In [71]:
# Carrega arquivo investimento_parte2
fin_02 = pd.read_csv('investimento_parte2.csv')

# Print do número de linhas e colunas do dataframe
print('Linhas', fin_02.shape[0])
print('Colunas', fin_02.shape[1])

# Primeiras cinco linhas do dataframe
fin_02.head()

Linhas 24052
Colunas 12


id    Salario       Perfil  Ativo01  Ativo02  Ativo03  Ativo04  \
0  id0000859  $51483.81    Agressivo  1273.95  5080.37  3565.44      NaN   
1  id0022202    $457.43  Conservador  1121.09  4927.51  3911.58   5511.7   
2  id0002056  $93245.46     Moderado      NaN  5283.71  2914.76      NaN   
3  id0005674   $3044.00    Agressivo  1353.76  5827.28  2998.24  16341.7   
4  id0004777    $880.30    Agressivo      NaN  5695.77  3525.98      NaN   

   Ativo05  Ativo06  Ativo07      Nota1      Nota2  
0  5635.72      NaN   513.38  47.599278   7.057028  
1  5593.40      NaN   497.28  37.353757  61.516665  
2  4707.33      NaN   499.06  98.895407  45.197866  
3  4712.65  2007.44   515.10  50.658089  65.660735  
4  4147.29      NaN   506.77  72.993882  55.530580

In [72]:
# Informações gerais sobre a base de dados cadastros

cadastro.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38754 entries, 0 to 38753
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        38754 non-null  object
 1   Nome      38754 non-null  object
 2   Sexo      38754 non-null  object
 3   Nasc      38754 non-null  object
 4   Idade     38754 non-null  int64 
 5   Email     38754 non-null  object
 6   Lat_long  38754 non-null  object
 7   Job       38754 non-null  object
 8   Estado    38754 non-null  object
dtypes: int64(1), object(8)
memory usage: 2.7+ MB


In [73]:
# Informações gerais sobre a base de dados investimento_1

fin_01.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       20000 non-null  object 
 1   Salario  20000 non-null  object 
 2   Perfil   20000 non-null  object 
 3   Ativo01  15012 non-null  float64
 4   Ativo02  16857 non-null  float64
 5   Ativo03  16760 non-null  float64
 6   Ativo04  13799 non-null  float64
 7   Ativo05  16829 non-null  float64
 8   Ativo06  9895 non-null   float64
 9   Ativo07  20000 non-null  float64
 10  Nota1    18469 non-null  float64
 11  Nota2    18954 non-null  float64
dtypes: float64(9), object(3)
memory usage: 1.8+ MB


In [74]:
# Informações gerais sobre a base de dados investimento_2

fin_02.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24052 entries, 0 to 24051
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       24052 non-null  object 
 1   Salario  24052 non-null  object 
 2   Perfil   24052 non-null  object 
 3   Ativo01  17994 non-null  float64
 4   Ativo02  20245 non-null  float64
 5   Ativo03  20249 non-null  float64
 6   Ativo04  16685 non-null  float64
 7   Ativo05  20211 non-null  float64
 8   Ativo06  11932 non-null  float64
 9   Ativo07  24052 non-null  float64
 10  Nota1    22172 non-null  float64
 11  Nota2    22810 non-null  float64
dtypes: float64(9), object(3)
memory usage: 2.2+ MB


In [75]:
# Algumas informações sobre os dados de cadastro
cadastro.describe().T

count       mean        std  min   25%   50%   75%   max
Idade  38754.0  25.716855  14.918449  0.0  13.0  26.0  39.0  51.0

In [76]:
cadastro.drop(axis=1, columns=['Lat_long']).describe(exclude=['int64']).T

count unique                   top   freq
id      38754  38754             id0000001      1
Nome    38754  38285          Ana da Silva      8
Sexo    38754      2                   Fem  19385
Nasc    38754  16399            1997-04-25     10
Email   38754  36008  jjohnson@hotmail.com      7
Job     38754    639         Tax inspector     84
Estado  38754     27                    PI   1503

In [77]:
# Algumas informações sobre os dados de investimento_1
fin_01.describe(exclude='object').T

count          mean          std          min          25%  \
Ativo01  15012.0   1125.757681   217.080495   800.070000   954.525000   
Ativo02  16857.0   5148.243110   398.701012  4500.090000  4833.260000   
Ativo03  16760.0   3139.165891   394.175941  2500.010000  2829.085000   
Ativo04  13799.0  15136.440041  6946.544888  5000.810000  9678.280000   
Ativo05  16829.0   5281.172070   792.562796  4000.140000  4654.050000   
Ativo06   9895.0   2317.491907   239.180603  2000.040000  2127.395000   
Ativo07  20000.0    500.093471    10.012402   462.930000   493.320000   
Nota1    18469.0     50.098208    28.952396     0.003143    25.020373   
Nota2    18954.0     50.174113    28.832285     0.007864    25.134177   

                  50%           75%           max  
Ativo01   1093.145000   1262.155000   2068.590000  
Ativo02   5107.230000   5407.630000   6817.340000  
Ativo03   3096.020000   3395.200000   5212.550000  
Ativo04  14070.350000  19424.205000  47467.730000  
Ativo05   5197.220000   5800.460000   9052.260000  
Ativo06   2270.870000   2455.335000   3610.050000  
Ativo07    500.130000    506.892500    542.050000  
Nota1       50.119452     75.238847     99.996659  
Nota2       50.242237     74.877627     99.991247

In [78]:
fin_01.describe(exclude='float').T

count unique        top  freq
id       20000  20000  id0011529     1
Salario  20000  16109      $3.50    11
Perfil   20000      3   Moderado  6692

In [79]:
# Algumas informações sobre os dados de investimento_2
fin_02.describe(exclude='object').T

count          mean          std          min          25%  \
Ativo01  17994.0   1129.080980   219.747037   800.010000   955.027500   
Ativo02  20245.0   5146.837462   398.977137  4500.050000  4831.890000   
Ativo03  20249.0   3141.053559   396.580165  2500.010000  2827.840000   
Ativo04  16685.0  15163.012785  6993.738051  5000.930000  9621.090000   
Ativo05  20211.0   5286.059975   796.892363  4000.250000  4654.300000   
Ativo06  11932.0   2320.358866   241.972481  2000.040000  2128.885000   
Ativo07  24052.0    500.019744     9.977632   462.930000   493.260000   
Nota1    22172.0     49.654722    28.913037     0.003143    24.767894   
Nota2    22810.0     49.973616    28.935261     0.003341    25.015850   

                  50%           75%           max  
Ativo01   1096.395000   1269.760000   2143.050000  
Ativo02   5101.630000   5409.750000   6990.260000  
Ativo03   3097.070000   3403.260000   4899.990000  
Ativo04  14090.430000  19517.340000  48744.740000  
Ativo05   5199.610000   5805.705000   9052.260000  
Ativo06   2270.135000   2464.085000   3639.650000  
Ativo07    500.120000    506.760000    542.300000  
Nota1       49.198346     74.880010     99.991343  
Nota2       49.747230     75.164789     99.999535

In [80]:
fin_02.describe(exclude='float').T

count unique        top  freq
id       24052  24052  id0000859     1
Salario  24052  19051      $3.08    14
Perfil   24052      3  Agressivo  8088

### 2 - Juntar os dataframes de investimento1 e investimento2 em um só e ver a quantidade de linhas

In [81]:
#Concatenação ou empilhamento vertical
financeiro = pd.concat([fin_01, fin_02])
financeiro

id    Salario       Perfil  Ativo01  Ativo02  Ativo03   Ativo04  \
0      id0011529      $2.25    Agressivo  1311.38      NaN      NaN  29267.48   
1      id0030644     $24.57    Agressivo  1007.38  4846.66  2717.82  12532.15   
2      id0019311    $367.53     Moderado  1128.99  4668.94  3092.60   9777.77   
3      id0036705     $23.26     Moderado  1185.26  4782.17  2669.05  14047.87   
4      id0001368      $6.31  Conservador   957.64  5585.52      NaN   5067.08   
...          ...        ...          ...      ...      ...      ...       ...   
24047  id0018222    $615.87  Conservador   933.65  5741.80  4031.59   6336.84   
24048  id0005515    $212.76  Conservador  1624.80      NaN  2689.68  21886.01   
24049  id0009893      $9.32    Agressivo   829.20      NaN  3831.45       NaN   
24050  id0016100  $81796.74  Conservador  1587.47  4881.66  4352.03  10391.18   
24051  id0027880  $38480.37    Agressivo   996.90  5604.43  2801.04   9183.62   

       Ativo05  Ativo06  Ativo07      Nota1      Nota2  
0      4993.30      NaN   506.40  73.007216  30.025005  
1      4282.98      NaN   499.93  20.000123  29.477728  
2      4246.30  2007.83   509.50  24.901450  49.659852  
3      6026.84  2064.09   489.82  14.246718  29.271206  
4      4030.51  2022.61   507.39   5.922253  11.840601  
...        ...      ...      ...        ...        ...  
24047      NaN      NaN   486.19        NaN  34.151046  
24048  4571.79      NaN   497.47  93.032164  93.621515  
24049  4784.34  2143.79   485.33   3.120521   9.086454  
24050  4695.44  2465.65   492.84  95.665966  30.164927  
24051  6220.96      NaN   503.24  55.035639  57.691996  

[44052 rows x 12 columns]

In [82]:
print('Linhas', financeiro.shape[0])
print('Colunas', financeiro.shape[1])

Linhas 44052
Colunas 12


### 3- Crie um novo dataframe eliminando as linhas duplicadas

In [83]:
# Aqui temos as linhas duplicadas do dataframe financeiro
colunas_duplicadas = financeiro[financeiro.duplicated()]
colunas_duplicadas.shape

(5298, 12)

In [84]:
# A função drop_duplicates() já remove as linhas duplicadas e retorna um dataframe sem as linhas duplicadas.
novo_fin = financeiro.drop_duplicates()
novo_fin

id    Salario       Perfil  Ativo01  Ativo02  Ativo03   Ativo04  \
0      id0011529      $2.25    Agressivo  1311.38      NaN      NaN  29267.48   
1      id0030644     $24.57    Agressivo  1007.38  4846.66  2717.82  12532.15   
2      id0019311    $367.53     Moderado  1128.99  4668.94  3092.60   9777.77   
3      id0036705     $23.26     Moderado  1185.26  4782.17  2669.05  14047.87   
4      id0001368      $6.31  Conservador   957.64  5585.52      NaN   5067.08   
...          ...        ...          ...      ...      ...      ...       ...   
24046  id0012328  $86268.66     Moderado   804.92  4520.13  3388.55  18396.16   
24048  id0005515    $212.76  Conservador  1624.80      NaN  2689.68  21886.01   
24049  id0009893      $9.32    Agressivo   829.20      NaN  3831.45       NaN   
24050  id0016100  $81796.74  Conservador  1587.47  4881.66  4352.03  10391.18   
24051  id0027880  $38480.37    Agressivo   996.90  5604.43  2801.04   9183.62   

       Ativo05  Ativo06  Ativo07      Nota1      Nota2  
0      4993.30      NaN   506.40  73.007216  30.025005  
1      4282.98      NaN   499.93  20.000123  29.477728  
2      4246.30  2007.83   509.50  24.901450  49.659852  
3      6026.84  2064.09   489.82  14.246718  29.271206  
4      4030.51  2022.61   507.39   5.922253  11.840601  
...        ...      ...      ...        ...        ...  
24046      NaN      NaN   504.90  97.685833  70.195608  
24048  4571.79      NaN   497.47  93.032164  93.621515  
24049  4784.34  2143.79   485.33   3.120521   9.086454  
24050  4695.44  2465.65   492.84  95.665966  30.164927  
24051  6220.96      NaN   503.24  55.035639  57.691996  

[38754 rows x 12 columns]

### 4 - Junte o dataframe de investimento com o de cadastro
Dica: use a chave id

In [85]:
# Com a função merge juntamos todas as informações, de ambas tabelas, fazemos um merge “outer”
# Podemos perceber que o número de linhas continua o mesmo mas o número de colunas aumentou.
dados_fin= pd.merge(cadastro, novo_fin, how = 'outer', on='id')
print('Linhas', dados_fin.shape[0])
print('Colunas', dados_fin.shape[1])
dados_fin.head()

Linhas 38754
Colunas 20


id                              Nome  Sexo        Nasc  Idade  \
0  id0000001             Eunice Hellen do Vale   Fem  1970-11-12     51   
1  id0000002   Sebastião Castelo Branco Júnior  Masc  2019-12-10      2   
2  id0000003         Romildo Fábio de Carvalho  Masc  2015-12-28      6   
3  id0000004         Lana Albina de Lemos Mumy   Fem  1993-05-03     28   
4  id0000005  Rose Elisângela Vyrnium de Moura   Fem  1984-06-10     37   

                     Email                                         Lat_long  \
0         djones@yahoo.com   (Decimal('24.7024355'), Decimal('-28.038258'))   
1   barnesandrea@yahoo.com   (Decimal('21.8589875'), Decimal('122.830566'))   
2  walkermichael@yahoo.com   (Decimal('-28.2161525'), Decimal('97.140529'))   
3   cobbangela@hotmail.com   (Decimal('-40.8557635'), Decimal('-2.936099'))   
4  crawfordcraig@yahoo.com  (Decimal('-20.2620885'), Decimal('104.349223'))   

                                     Job Estado    Salario       Perfil  \
0                       Technical author     TO     $63.67     Moderado   
1                   Biochemist, clinical     DF     $21.96     Moderado   
2  Development worker, international aid     DF  $62497.64    Agressivo   
3              Estate manager/land agent     DF  $99878.74    Agressivo   
4    Chartered public finance accountant     MS    $188.01  Conservador   

   Ativo01  Ativo02  Ativo03   Ativo04  Ativo05  Ativo06  Ativo07      Nota1  \
0  1174.27  4980.98  2638.24       NaN  4480.91  2329.72   480.72  79.600372   
1   842.12  5646.09  3336.07       NaN  5624.22  2669.98   504.17  44.406893   
2   818.81  5516.52  3396.64       NaN  5033.83  2229.45   523.12  44.020024   
3      NaN  5258.69  3723.10       NaN  5933.91      NaN   486.05  25.597924   
4  1505.16  6415.30  3462.67  30841.35  6044.86  2198.89   512.98  14.463307   

       Nota2  
0   9.802620  
1  90.512871  
2  65.060941  
3  86.548918  
4  33.566292

### 5 - Ordene o dataframe pelo id, de forma crescente e redefina o índice

In [86]:
# Classifica o df pela coluna id
dados_fin.sort_values(by='id')

# Cria um novo index e remove a coluna do index antigo
dados_fin = dados_fin.reset_index().drop(['index'], axis=1)

dados_fin

id                                Nome  Sexo        Nasc  Idade  \
0      id0000001               Eunice Hellen do Vale   Fem  1970-11-12     51   
1      id0000002     Sebastião Castelo Branco Júnior  Masc  2019-12-10      2   
2      id0000003           Romildo Fábio de Carvalho  Masc  2015-12-28      6   
3      id0000004           Lana Albina de Lemos Mumy   Fem  1993-05-03     28   
4      id0000005    Rose Elisângela Vyrnium de Moura   Fem  1984-06-10     37   
...          ...                                 ...   ...         ...    ...   
38749  id0038750     Beatriz Maísa de Miranda Milani   Fem  1999-02-20     22   
38750  id0038751                     Juan dos Santos  Masc  1999-07-10     22   
38751  id0038752       Severino Edson de Viana Loase  Masc  1981-12-01     40   
38752  id0038753                          Kim Mooshi  Masc  1999-12-06     22   
38753  id0038754  Vinícius Arnaldo de Pereira Mendes  Masc  2003-09-13     18   

                         Email  \
0             djones@yahoo.com   
1       barnesandrea@yahoo.com   
2      walkermichael@yahoo.com   
3       cobbangela@hotmail.com   
4      crawfordcraig@yahoo.com   
...                        ...   
38749         mark47@gmail.com   
38750  greggnguyen@hotmail.com   
38751        cindy11@yahoo.com   
38752  mcdonaldmario@gmail.com   
38753    omartinez@hotmail.com   

                                              Lat_long  \
0       (Decimal('24.7024355'), Decimal('-28.038258'))   
1       (Decimal('21.8589875'), Decimal('122.830566'))   
2       (Decimal('-28.2161525'), Decimal('97.140529'))   
3       (Decimal('-40.8557635'), Decimal('-2.936099'))   
4      (Decimal('-20.2620885'), Decimal('104.349223'))   
...                                                ...   
38749  (Decimal('-23.8494875'), Decimal('-91.561304'))   
38750   (Decimal('-7.8876355'), Decimal('155.835880'))   
38751  (Decimal('-52.1338005'), Decimal('-59.715325'))   
38752  (Decimal('-81.2874995'), Decimal('-28.805800'))   
38753     (Decimal('-37.248521'), Decimal('6.033254'))   

                                         Job Estado    Salario       Perfil  \
0                           Technical author     TO     $63.67     Moderado   
1                       Biochemist, clinical     DF     $21.96     Moderado   
2      Development worker, international aid     DF  $62497.64    Agressivo   
3                  Estate manager/land agent     DF  $99878.74    Agressivo   
4        Chartered public finance accountant     MS    $188.01  Conservador   
...                                      ...    ...        ...          ...   
38749             Further education lecturer     RS   $4993.49     Moderado   
38750                  Adult guidance worker     PI  $41218.63     Moderado   
38751               Arts development officer     PR     $32.41    Agressivo   
38752                     Production manager     RJ    $914.52  Conservador   
38753                 Pension scheme manager     AL  $29076.19     Moderado   

       Ativo01  Ativo02  Ativo03   Ativo04  Ativo05  Ativo06  Ativo07  \
0      1174.27  4980.98  2638.24       NaN  4480.91  2329.72   480.72   
1       842.12  5646.09  3336.07       NaN  5624.22  2669.98   504.17   
2       818.81  5516.52  3396.64       NaN  5033.83  2229.45   523.12   
3          NaN  5258.69  3723.10       NaN  5933.91      NaN   486.05   
4      1505.16  6415.30  3462.67  30841.35  6044.86  2198.89   512.98   
...        ...      ...      ...       ...      ...      ...      ...   
38749  1001.01  5561.55      NaN   8454.99      NaN  2191.78   489.76   
38750  1532.51  5102.84  3097.50  11243.08  4692.11  2368.93   496.02   
38751  1163.90  4911.63  3746.62       NaN  6733.79  2338.43   504.32   
38752      NaN  5532.35  2861.32       NaN  4921.77  2024.00   500.65   
38753   981.62  4825.10      NaN  18085.88  5161.07  2257.66   507.95   

           Nota1      Nota2  
0      79.600372   9.802620  
1      44.406893  90.512871  
2      44.020024  6

In [87]:
# Ordena de forma crescente
dados_fin.sort_index(ascending=True)

id                                Nome  Sexo        Nasc  Idade  \
0      id0000001               Eunice Hellen do Vale   Fem  1970-11-12     51   
1      id0000002     Sebastião Castelo Branco Júnior  Masc  2019-12-10      2   
2      id0000003           Romildo Fábio de Carvalho  Masc  2015-12-28      6   
3      id0000004           Lana Albina de Lemos Mumy   Fem  1993-05-03     28   
4      id0000005    Rose Elisângela Vyrnium de Moura   Fem  1984-06-10     37   
...          ...                                 ...   ...         ...    ...   
38749  id0038750     Beatriz Maísa de Miranda Milani   Fem  1999-02-20     22   
38750  id0038751                     Juan dos Santos  Masc  1999-07-10     22   
38751  id0038752       Severino Edson de Viana Loase  Masc  1981-12-01     40   
38752  id0038753                          Kim Mooshi  Masc  1999-12-06     22   
38753  id0038754  Vinícius Arnaldo de Pereira Mendes  Masc  2003-09-13     18   

                         Email  \
0             djones@yahoo.com   
1       barnesandrea@yahoo.com   
2      walkermichael@yahoo.com   
3       cobbangela@hotmail.com   
4      crawfordcraig@yahoo.com   
...                        ...   
38749         mark47@gmail.com   
38750  greggnguyen@hotmail.com   
38751        cindy11@yahoo.com   
38752  mcdonaldmario@gmail.com   
38753    omartinez@hotmail.com   

                                              Lat_long  \
0       (Decimal('24.7024355'), Decimal('-28.038258'))   
1       (Decimal('21.8589875'), Decimal('122.830566'))   
2       (Decimal('-28.2161525'), Decimal('97.140529'))   
3       (Decimal('-40.8557635'), Decimal('-2.936099'))   
4      (Decimal('-20.2620885'), Decimal('104.349223'))   
...                                                ...   
38749  (Decimal('-23.8494875'), Decimal('-91.561304'))   
38750   (Decimal('-7.8876355'), Decimal('155.835880'))   
38751  (Decimal('-52.1338005'), Decimal('-59.715325'))   
38752  (Decimal('-81.2874995'), Decimal('-28.805800'))   
38753     (Decimal('-37.248521'), Decimal('6.033254'))   

                                         Job Estado    Salario       Perfil  \
0                           Technical author     TO     $63.67     Moderado   
1                       Biochemist, clinical     DF     $21.96     Moderado   
2      Development worker, international aid     DF  $62497.64    Agressivo   
3                  Estate manager/land agent     DF  $99878.74    Agressivo   
4        Chartered public finance accountant     MS    $188.01  Conservador   
...                                      ...    ...        ...          ...   
38749             Further education lecturer     RS   $4993.49     Moderado   
38750                  Adult guidance worker     PI  $41218.63     Moderado   
38751               Arts development officer     PR     $32.41    Agressivo   
38752                     Production manager     RJ    $914.52  Conservador   
38753                 Pension scheme manager     AL  $29076.19     Moderado   

       Ativo01  Ativo02  Ativo03   Ativo04  Ativo05  Ativo06  Ativo07  \
0      1174.27  4980.98  2638.24       NaN  4480.91  2329.72   480.72   
1       842.12  5646.09  3336.07       NaN  5624.22  2669.98   504.17   
2       818.81  5516.52  3396.64       NaN  5033.83  2229.45   523.12   
3          NaN  5258.69  3723.10       NaN  5933.91      NaN   486.05   
4      1505.16  6415.30  3462.67  30841.35  6044.86  2198.89   512.98   
...        ...      ...      ...       ...      ...      ...      ...   
38749  1001.01  5561.55      NaN   8454.99      NaN  2191.78   489.76   
38750  1532.51  5102.84  3097.50  11243.08  4692.11  2368.93   496.02   
38751  1163.90  4911.63  3746.62       NaN  6733.79  2338.43   504.32   
38752      NaN  5532.35  2861.32       NaN  4921.77  2024.00   500.65   
38753   981.62  4825.10      NaN  18085.88  5161.07  2257.66   507.95   

           Nota1      Nota2  
0      79.600372   9.802620  
1      44.406893  90.512871  
2      44.020024  6

## 6 - Faça uma análise dos dados faltantes e trate-os

Mostre todas as colunas com dados faltantes, sua quantidade/porcentagem e como tratar (preencher) cada uma delas explicando sua estratégia


In [88]:
#Soma dos valores faltantes de todas as colunas
dados_fin.isnull().sum()



id              0
Nome            0
Sexo            0
Nasc            0
Idade           0
Email           0
Lat_long        0
Job             0
Estado          0
Salario         0
Perfil          0
Ativo01      9724
Ativo02      6100
Ativo03      6169
Ativo04     11887
Ativo05      6191
Ativo06     19514
Ativo07         0
Nota1        3000
Nota2        2000
dtype: int64

#### Como temos 38754 linhas de informação, caso fossemos excluir todas as linhas em que contém dados faltantes teríamos ao final apenas 5147 linhas. Como pode ser visto no exemplo abaixo:


In [89]:
#Cria um dataframe de teste
dados_teste=dados_fin.copy()
#Verifica números de linhas e colunas do dataframe de teste
print('Linhas', dados_teste.shape[0])
print('Colunas', dados_teste.shape[1])

Linhas 38754
Colunas 20


In [90]:
#Exclue todos as linhas com dados faltantes
dados_teste.dropna(inplace=True)
#Verifica números de linhas e colunas do dataframe de teste
print('Linhas', dados_teste.shape[0])
print('Colunas', dados_teste.shape[1])

Linhas 5147
Colunas 20


### Começando o tratamento de dados faltantes para todas as colunas de Ativos, podemos analisar os valores mais frequentes com os conceitos de média, moda e mediana.

In [91]:
# Trazendo média, moda e mediana dos dados da coluna Ativo06
print(f'A média dos dados do Ativo 06 é:{dados_fin.Ativo06.mean()}')
print(f'A moda dos dados do Ativo 06 é:{dados_fin.Ativo06.mode()[0]}')
print(f'A mediana dos dados do Ativo 06 é:{dados_fin.Ativo06.median()}')

A média dos dados do Ativo 06 é:2319.482555093555
A moda dos dados do Ativo 06 é:2030.03
A mediana dos dados do Ativo 06 é:2271.385


#### Como a média é maior que a mediana e a mediana é maior do que a moda, os dados são assimétricos e dessa forma a mediana é um valor mais fiel a tendência dos dados do que a média. 

In [93]:
#Extraindo a mediana da coluna Ativo06
mediana=dados_fin.Ativo06.median()
#Preenchendo os valores faltantes com a mediana
dados_fin.Ativo06.fillna(mediana, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id              0
Nome            0
Sexo            0
Nasc            0
Idade           0
Email           0
Lat_long        0
Job             0
Estado          0
Salario         0
Perfil          0
Ativo01      9724
Ativo02      6100
Ativo03      6169
Ativo04     11887
Ativo05      6191
Ativo06         0
Ativo07         0
Nota1        3000
Nota2        2000
dtype: int64

### Segue o mesmo padrão para todos os outros ativos.

In [94]:
# Trazendo média, moda e mediana dos dados da coluna Ativo05
print(f'A média dos dados do Ativo 05 é:{dados_fin.Ativo05.mean()}')
print(f'A moda dos dados do Ativo 05 é:{dados_fin.Ativo05.mode()[0]}')
print(f'A mediana dos dados do Ativo 05 é:{dados_fin.Ativo05.median()}')

A média dos dados do Ativo 05 é:5284.1206614869625
A moda dos dados do Ativo 05 é:4120.89
A mediana dos dados do Ativo 05 é:5198.42


In [96]:
#Extraindo a mediana da coluna Ativo05
mediana=dados_fin.Ativo05.median()
#Preenchendo os valores faltantes com a mediana
dados_fin.Ativo05.fillna(mediana, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id              0
Nome            0
Sexo            0
Nasc            0
Idade           0
Email           0
Lat_long        0
Job             0
Estado          0
Salario         0
Perfil          0
Ativo01      9724
Ativo02      6100
Ativo03      6169
Ativo04     11887
Ativo05         0
Ativo06         0
Ativo07         0
Nota1        3000
Nota2        2000
dtype: int64

In [97]:
# Trazendo média, moda e mediana dos dados da coluna Ativo04
print(f'A média dos dados do Ativo 04 é:{dados_fin.Ativo04.mean()}')
print(f'A moda dos dados do Ativo 04 é:{dados_fin.Ativo04.mode()[0]}')
print(f'A mediana dos dados do Ativo 04 é:{dados_fin.Ativo04.median()}')

A média dos dados do Ativo 04 é:15153.68174154167
A moda dos dados do Ativo 04 é:13810.37
A mediana dos dados do Ativo 04 é:14066.76


In [98]:
#Extraindo a mediana da coluna Ativo04
mediana=dados_fin.Ativo04.median()
#Preenchendo os valores faltantes com a mediana
dados_fin.Ativo04.fillna(mediana, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id             0
Nome           0
Sexo           0
Nasc           0
Idade          0
Email          0
Lat_long       0
Job            0
Estado         0
Salario        0
Perfil         0
Ativo01     9724
Ativo02     6100
Ativo03     6169
Ativo04        0
Ativo05        0
Ativo06        0
Ativo07        0
Nota1       3000
Nota2       2000
dtype: int64

In [99]:
# Trazendo média, moda e mediana dos dados da coluna Ativo03
print(f'A média dos dados do Ativo 03 é:{dados_fin.Ativo03.mean()}')
print(f'A moda dos dados do Ativo 03 é:{dados_fin.Ativo03.mode()[0]}')
print(f'A mediana dos dados do Ativo 03 é:{dados_fin.Ativo03.median()}')

A média dos dados do Ativo 03 é:3140.4189307963793
A moda dos dados do Ativo 03 é:3046.22
A mediana dos dados do Ativo 03 é:3096.08


In [100]:
#Extraindo a mediana da coluna Ativo03
mediana=dados_fin.Ativo03.median()
#Preenchendo os valores faltantes com a mediana
dados_fin.Ativo03.fillna(mediana, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id             0
Nome           0
Sexo           0
Nasc           0
Idade          0
Email          0
Lat_long       0
Job            0
Estado         0
Salario        0
Perfil         0
Ativo01     9724
Ativo02     6100
Ativo03        0
Ativo04        0
Ativo05        0
Ativo06        0
Ativo07        0
Nota1       3000
Nota2       2000
dtype: int64

In [101]:
# Trazendo média, moda e mediana dos dados da coluna Ativo02
print(f'A média dos dados do Ativo 02 é:{dados_fin.Ativo02.mean()}')
print(f'A moda dos dados do Ativo 02 é:{dados_fin.Ativo02.mode()[0]}')
print(f'A mediana dos dados do Ativo 02 é:{dados_fin.Ativo02.median()}')

A média dos dados do Ativo 02 é:5147.296097874686
A moda dos dados do Ativo 02 é:4666.63
A mediana dos dados do Ativo 02 é:5103.425


In [102]:
#Extraindo a mediana da coluna Ativo02
mediana=dados_fin.Ativo02.median()
#Preenchendo os valores faltantes com a mediana
dados_fin.Ativo02.fillna(mediana, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id             0
Nome           0
Sexo           0
Nasc           0
Idade          0
Email          0
Lat_long       0
Job            0
Estado         0
Salario        0
Perfil         0
Ativo01     9724
Ativo02        0
Ativo03        0
Ativo04        0
Ativo05        0
Ativo06        0
Ativo07        0
Nota1       3000
Nota2       2000
dtype: int64

In [103]:
# Trazendo média, moda e mediana dos dados da coluna Ativo01
print(f'A média dos dados do Ativo 01 é:{dados_fin.Ativo01.mean()}')
print(f'A moda dos dados do Ativo 01 é:{dados_fin.Ativo01.mode()[0]}')
print(f'A mediana dos dados do Ativo 01 é:{dados_fin.Ativo01.median()}')

A média dos dados do Ativo 01 é:1128.3393003789183
A moda dos dados do Ativo 01 é:1091.55
A mediana dos dados do Ativo 01 é:1095.51


In [104]:
#Extraindo a mediana da coluna Ativo01
mediana=dados_fin.Ativo01.median()
#Preenchendo os valores faltantes com a mediana
dados_fin.Ativo01.fillna(mediana, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id             0
Nome           0
Sexo           0
Nasc           0
Idade          0
Email          0
Lat_long       0
Job            0
Estado         0
Salario        0
Perfil         0
Ativo01        0
Ativo02        0
Ativo03        0
Ativo04        0
Ativo05        0
Ativo06        0
Ativo07        0
Nota1       3000
Nota2       2000
dtype: int64

### Já para o tratamento de dados faltantes nas colunas de Notas, percebemos um padrão diferente ao analisar a média, moda e mediana.

In [106]:
# Trazendo média, moda e mediana dos dados da coluna Nota1
print(f'A média dos dados do Nota 1 é:{dados_fin.Nota1.mean()}')
print(f'A moda dos dados do Nota 1 é:{dados_fin.Nota1.mode()[0]}')
print(f'A mediana dos dados do Nota 1 é:{dados_fin.Nota1.median()}')

A média dos dados do Nota 1 é:49.82109426961431
A moda dos dados do Nota 1 é:0.0031431914018997
A mediana dos dados do Nota 1 é:49.584429781574606


#### Como a média é semelhante com a mediana, os dados são simetricos e dessa forma a media é um valor mais fiel a tendência dos dados.

In [107]:
#Extraindo a mediana da coluna Nota1
media=dados_fin.Nota1.mean()
#Preenchendo os valores faltantes com a mediana
dados_fin.Nota1.fillna(media, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id             0
Nome           0
Sexo           0
Nasc           0
Idade          0
Email          0
Lat_long       0
Job            0
Estado         0
Salario        0
Perfil         0
Ativo01        0
Ativo02        0
Ativo03        0
Ativo04        0
Ativo05        0
Ativo06        0
Ativo07        0
Nota1          0
Nota2       2000
dtype: int64

In [108]:
# Trazendo média, moda e mediana dos dados da coluna Nota2
print(f'A média dos dados do Nota 2 é:{dados_fin.Nota2.mean()}')
print(f'A moda dos dados do Nota 2 é:{dados_fin.Nota2.mode()[0]}')
print(f'A mediana dos dados do Nota 2 é:{dados_fin.Nota2.median()}')

A média dos dados do Nota 2 é:50.075481092463434
A moda dos dados do Nota 2 é:0.0033411307592712
A mediana dos dados do Nota 2 é:50.06459173080374


In [109]:
#Extraindo a mediana da coluna Nota2
media=dados_fin.Nota2.mean()
#Preenchendo os valores faltantes com a mediana
dados_fin.Nota2.fillna(media, inplace=True)
#Checa se os valores foram mudados
dados_fin.isnull().sum()

id          0
Nome        0
Sexo        0
Nasc        0
Idade       0
Email       0
Lat_long    0
Job         0
Estado      0
Salario     0
Perfil      0
Ativo01     0
Ativo02     0
Ativo03     0
Ativo04     0
Ativo05     0
Ativo06     0
Ativo07     0
Nota1       0
Nota2       0
dtype: int64